bin/proving-service/src/proxy/worker.rs

```diff
@@ -1,75 +1,216 @@
 
         Ok(Self {
-            backend: worker,
+            backend,
             is_available: true,
-            health_check_client,
+            status_client,
+            health_status: WorkerHealthStatus::Healthy,
+            version: "".to_string(),
         })
     }
 
+    /// Returns the worker address.
     pub fn address(&self) -> String {
         self.backend.addr.to_string()
     }
 
-    pub async fn is_healthy(&mut self) -> bool {
-        match self
-            .health_check_client
-            .check(HealthCheckRequest { service: "".to_string() })
-            .await
-        {
-            Ok(response) => response.into_inner().status() == ServingStatus::Serving,
-            Err(err) => {
-                error!("Failed to check worker health ({}): {}", self.address(), err);
-                false
+    /// Checks the worker status.
+    ///
+    /// # Returns
+    /// - `Some(true)` if the worker is ready.
+    /// - `Some(false)` if the worker is not ready or if there was an error checking the status.
+    /// - `None` if the worker should not do a health check.
+    pub async fn check_status(
+        &mut self,
+        supported_prover_type: &ProverType,
+    ) -> Option<WorkerHealthStatus> {
+        if !self.should_do_health_check() {
+            return None;
+        }
+
+        let failed_attempts = self.retries_amount();
+
+        let worker_status = match self.status_client.status(StatusRequest {}).await {
+            Ok(response) => response.into_inner(),
+            Err(e) => {
+                error!("Failed to check worker status ({}): {}", self.address(), e);
+                return Some(WorkerHealthStatus::Unhealthy {
+                    failed_attempts: failed_attempts + 1,
+                    first_fail_timestamp: Instant::now(),
+                    reason: e.message().to_string(),
+                });
             },
+        };
+
+        self.version = worker_status.version;
+
+        let worker_supported_proof_type =
+            match ProverType::try_from(worker_status.supported_proof_type) {
+                Ok(proof_type) => proof_type,
+                Err(e) => {
+                    error!(
+                        "Failed to convert worker supported proof type ({}): {}",
+                        self.address(),
+                        e
+                    );
+                    return Some(WorkerHealthStatus::Unhealthy {
+                        failed_attempts: failed_attempts + 1,
+                        first_fail_timestamp: Instant::now(),
+                        reason: e.to_string(),
+                    });
+                },
+            };
+
+        if !(*supported_prover_type == worker_supported_proof_type) {
+            return Some(WorkerHealthStatus::Unhealthy {
+                failed_attempts: failed_attempts + 1,
+                first_fail_timestamp: Instant::now(),
+                reason: "Unsupported proof type".to_string(),
```

Comment:
It would be nice if we could add what proof type the worker supports to help in the troubleshooting process

Reply:
Done!